proofs/proof.ml

```diff
@@ -366,33 +363,29 @@ let update_sigma_env p env =
 
 let run_tactic env tac pr =
   let open Proofview.Notations in
-  let sp = pr.proofview in
   let undef sigma l = List.filter (fun g -> Evd.is_undefined sigma g) l in
   let tac =
     tac >>= fun result ->
     Proofview.tclEVARMAP >>= fun sigma ->
     (* Already solved goals are not to be counted as shelved. Nor are
       they to be marked as unresolvable. *)
-    let retrieved = Evd.filter_future_goals (Evd.is_undefined sigma) (Evd.save_future_goals sigma) in
-    let retrieved,retrieved_given_up = Evd.extract_given_up_future_goals retrieved in
-    (* Check that retrieved given up is empty *)
-    if not (List.is_empty retrieved_given_up) then
-      CErrors.anomaly Pp.(str "Evars generated outside of proof engine (e.g. V82, clear, ...) are not supposed to be explicitly given up.");
+    let retrieved, sigma = Evd.pop_future_goals sigma in
+    let retrieved = Evd.FutureGoals.filter (Evd.is_undefined sigma) retrieved in
```

Comment:
Not different from before but, why is this simply "undefined" and not up-to advance? 
If a future goal is instantiated by another new evar, I guess the other evar should be a future goal too. It could also be an older goal, and we don't want to consider this possibility here. Is that reasoning correct?

Reply:
Aren't all new evars (until they are instantiated) declared as future goals?